From 527761f0a083172c6e257027b93571cce8568acf Mon Sep 17 00:00:00 2001
From: hayao <shun819.mail@gmail.com>
Date: Mon, 27 Jul 2026 18:35:14 +0900
Subject: [PATCH 05/16] x86/pc9800: put the 8253 PIT at its PC-9800 ports and
 clock

Counters at 0x71/0x73/0x75, control at 0x77 (not 0x40-0x43), and a 2.4576 MHz
input clock (NP2kai's default 5MHz-family machine) instead of the PC/AT's
1.193182 MHz. With this the PIT clockevent actually ticks: the kernel
calibrates, switches to the pit clocksource and boots to the init search.
---
 include/linux/i8253.h | 7 +++++++
 include/linux/timex.h | 6 ++++++
 2 files changed, 13 insertions(+)

diff --git a/include/linux/i8253.h b/include/linux/i8253.h
index 56c280eb2d4f..7c47fb06c072 100644
--- a/include/linux/i8253.h
+++ b/include/linux/i8253.h
@@ -14,9 +14,16 @@
 #include <linux/timex.h>
 
 /* i8253A PIT registers */
+#ifdef CONFIG_X86_PC9800
+/* NEC PC-9800: 8253 counters at I/O 0x71/0x73/0x75, control at 0x77. */
+#define PIT_MODE	0x77
+#define PIT_CH0		0x71
+#define PIT_CH2		0x75
+#else
 #define PIT_MODE	0x43
 #define PIT_CH0		0x40
 #define PIT_CH2		0x42
+#endif
 
 #define PIT_LATCH	((PIT_TICK_RATE + HZ/2) / HZ)
 
diff --git a/include/linux/timex.h b/include/linux/timex.h
index 3871b06bd302..ea841bd798d8 100644
--- a/include/linux/timex.h
+++ b/include/linux/timex.h
@@ -167,6 +167,12 @@ extern void hardpps(const struct timespec64 *, const struct timespec64 *);
 int read_current_timer(unsigned long *timer_val);
 
 /* The clock frequency of the i8253/i8254 PIT */
+#ifdef CONFIG_X86_PC9800
+/* PC-9800's 8253 is clocked at 2.4576 MHz on 5MHz-family machines (NP2kai's
+ * default); 8MHz-family machines use 1.9968 MHz. */
+#define PIT_TICK_RATE 2457600ul
+#else
 #define PIT_TICK_RATE 1193182ul
+#endif
 
 #endif /* LINUX_TIMEX_H */
-- 
2.54.0

